File: src/SharpOMatic.Engine/Contexts/AssetRef.cs
namespace SharpOMatic.Engine.Contexts;

public class AssetRef
{
    public AssetRef() { }

    [SetsRequiredMembers]
    public AssetRef(Asset asset)
    {
        if (asset is null)
            throw new ArgumentNullException(nameof(asset));

        AssetId = asset.AssetId;
        Name = asset.Name;
        MediaType = asset.MediaType;
        SizeBytes = asset.SizeBytes;
        FolderId = asset.FolderId;
    }

    public required Guid AssetId { get; set; }
    public required string Name { get; set; }
    public required string MediaType { get; set; }
    public required long SizeBytes { get; set; }
    public Guid? FolderId { get; set; }
    public string? FolderName { get; set; }
}

File: src/SharpOMatic.FrontEnd/src/app/entities/definitions/asset-ref.ts
export interface AssetRef {
  assetId: string;
  name: string;
  mediaType: string;
  sizeBytes: number;
  folderId?: string | null;
  folderName?: string | null;
}

const isAssetRef = (value: unknown): value is AssetRef => {
  if (!value || typeof value !== 'object') {
    return false;
  }

  const candidate = value as AssetRef;
  return (
    typeof candidate.assetId === 'string' &&
    typeof candidate.name === 'string' &&
    typeof candidate.mediaType === 'string' &&
    typeof candidate.sizeBytes === 'number' &&
    (candidate.folderId === undefined ||
      candidate.folderId === null ||
      typeof candidate.folderId === 'string') &&
    (candidate.folderName === undefined ||
      candidate.folderName === null ||
      typeof candidate.folderName === 'string')
  );
};

export const parseAssetRefValue = (value: string): AssetRef | null => {
  if (!value) {
    return null;
  }

  try {
    const parsed = JSON.parse(value);
    return isAssetRef(parsed) ? parsed : null;
  } catch {
    return null;
  }
};

export const parseAssetRefListValue = (value: string): AssetRef[] => {
  if (!value) {
    return [];
  }

  try {
    const parsed = JSON.parse(value);
    if (!Array.isArray(parsed)) {
      return [];
    }

    return parsed.filter(isAssetRef);
  } catch {
    return [];
  }
};

export const buildAssetRefValue = (asset: AssetRef | null): string => {
  return asset ? JSON.stringify(asset) : '';
};

export const buildAssetRefListValue = (assets: AssetRef[]): string => {
  return JSON.stringify(assets);
};

export const formatAssetRefLabel = (asset: AssetRef): string => {
  if (asset.folderName) {
    return `${asset.folderName}/${asset.name}`;
  }

  return asset.name;
};
